copybookreader-plugins/src/main/java/co/cask/hydrator/plugin/batch/CopyBookSource.java

```diff
@@ -0,0 +1,291 @@
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.batch.BatchSource;
+import co.cask.cdap.etl.api.batch.BatchSourceContext;
+import co.cask.hydrator.common.ReferencePluginConfig;
+import co.cask.hydrator.common.SourceInputFormatProvider;
+import co.cask.hydrator.common.batch.JobUtils;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.collect.Lists;
+import org.apache.avro.reflect.Nullable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.mapreduce.Job;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Batch Source to poll fixed length flat files that can be parsed using CobolCopybook.
+ * The plugin will accept the Copybook contents in a textbox and a bianry data file.
+ * <p>
+ * It produces structured record based on the schema as defined either by the Copybook contents or as defined by user
+ * <p>
+ * For the first implementation it will only accept binary fixed length flat files without any nesting
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name("CopybookReader")
+@Description("Batch Source to read COBOL Copybook fixed length flat files")
+public class CopybookSource extends BatchSource<LongWritable, Map<String, String>, StructuredRecord> {
+
+  @Description("If no file type is mentioned, by default it will be considered as a fixed length flat file")
+  public static final int DEFAULT_FILE_STRUCTURE = net.sf.JRecord.Common.Constants.IO_FIXED_LENGTH;
+
+  @Description("Copybook file type supported. For the current implementation the Copybook reader will accept only " +
+    "fixed flat files")
+  public static final List<Integer> SUPPORTED_FILE_STRUCTURES = Lists.newArrayList(
+    net.sf.JRecord.Common.Constants.IO_FIXED_LENGTH
+  );
+  public static final String COPYBOOK_INPUTFORMAT_CBL_CONTENTS = "copybook.inputformat.cbl.contents";
+  /* For the initial implementation only fixed length binary files will be accepted.
+   This will not handle complex nested structures of COBOL copybook
+   Also this will not handle Redefines or iterators in the structure*/
+  public static final String COPYBOOK_INPUTFORMAT_FILE_STRUCTURE = "copybook.inputformat.input.filestructure";
+  public static final String COPYBOOK_INPUTFORMAT_DATA_HDFS_PATH = "copybook.inputformat.data.hdfs.path";
+
+  private final CopybookSourceConfig config;
+
+  public CopybookSource(CopybookSourceConfig copybookConfig) {
+    this.config = copybookConfig;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validate();
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(config.parseSchema());
```

Comment:
Yes , JRecord has provided types, but those are different than the Java data types. Hence need to do required conversion in CopybookReader class. However, not sure of the exact mapping of the data types and will update code after discussion with Sree.

Also, as per requirement user can select specific fields from the copybook contents as output schema, because the user may not need all the fields. Hence schema field is provided on the plugin for user specific schema.